lucene/core/src/java/org/apache/lucene/search/DoubleValuesSource.java

```diff
@@ -617,8 +618,17 @@ public double doubleValue() throws IOException {
         public boolean advanceExact(int doc) throws IOException {
           if (disi.docID() < doc) {
             disi.advance(doc);
+            tpiMatch = null;
           }
-          return disi.docID() == doc && (tpi == null || tpi.matches());
+          if (disi.docID() == doc) {
```

Comment:
Rather than a `Boolean`, how about this?
```
private boolean matched = false; // avoid calling tpi.matches() multiple times on a single doc
...
public boolean advanceExact(int doc) throws IOException {
  if (disi.docId() == doc) {
    return matched;
  }
  if (disi.docId() < doc) {
    disi.advance(doc);
  }
  return matched = disi.docId() == doc && (tpi == null || tpi.matches());
}
```

Reply:
At first glance, this looked simpler and correct, but on further reflection, I think this contains a bug.  Lets say advanceExact doc==1 and we call disi.advance(1) and it arrives at 2 because the disi doesn't match 1.  We'd then set matched=false and return (fine; it doesn't match doc==1).  Then say the caller calls advanceExact doc==2.  We'd quick return "matched" which is false.  But that's a mistake because the disi in fact matches doc==2 (ignoring tpi; say it's null).

I like the Boolean tpiMatch because it caches the very thing that needs to be cached, and it holds null to say we don't know or it doesn't matter.

Reply:
Good catch.  Let's go with the Boolean then.